bindings/rust/examples/client-hello-config-resolution/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+    configs: HashMap<String, s2n_tls::config::Config>,
+}
+
+impl AnimalConfigResolver {
+    pub fn new() -> Self {
+        let mut configs = HashMap::new();
+        configs.insert("www.wombat.com".to_owned(), server_config("wombat"));
+        configs.insert("www.kangaroo.com".to_owned(), server_config("kangaroo"));
+        Self { configs }
+    }
+}
+
+impl ClientHelloCallback for AnimalConfigResolver {
+    fn on_client_hello(
+        &self,
+        connection: &mut s2n_tls::connection::Connection,
+    ) -> Result<Option<Pin<Box<dyn ConnectionFuture>>>, s2n_tls::error::Error> {
+        let sni = match connection.server_name() {
+            Some(sni) => sni,
+            None => {
+                return Err(s2n_tls::error::Error::application(
+                    "connection contained no sni".into(),
+                ))
+            }
+        };
+        let config_ref = match self.configs.get(sni) {
+            Some(c) => c,
+            None => {
+                return Err(s2n_tls::error::Error::application(
+                    format!("unexpected SNI {sni}").into(),
+                ))
+            }
+        };
+        println!("client hello callback: setting connection config associated with {sni}");
+        let config = config_ref.clone();
+        connection.set_config(config).unwrap();
+        // Ok -> the function completed successfully
+        // None -> s2n-tls doesn't need to poll this to completion
+        Ok(None)
+    }
+}
+
+fn server_config(animal: &str) -> s2n_tls::config::Config {
+    let cert_path = format!("{}/certs/{}-chain.pem", env!("CARGO_MANIFEST_DIR"), animal);
+    let key_path = format!("{}/certs/{}-key.pem", env!("CARGO_MANIFEST_DIR"), animal);
+    let cert = std::fs::read(cert_path).unwrap();
+    let key = std::fs::read(key_path).unwrap();
+    let mut config = s2n_tls::config::Builder::new();
+
+    // we can set different policies for different configs. "20190214" doesn't
+    // support TLS 1.3, so any customer requesting www.wombat.com won't be able
+    // to negotiate TLS 1.3
+    let security_policy = match animal {
+        "wombat" => Policy::from_version("20190214").unwrap(),
+        _ => DEFAULT_TLS13,
+    };
+    config.set_security_policy(&security_policy).unwrap();
+    config.load_pem(&cert, &key).unwrap();
+    config.build().unwrap()
+}
+
+pub fn client_config() -> s2n_tls::config::Config {
+    let mut config = s2n_tls::config::Config::builder();
+    let ca: Vec<u8> =
+        std::fs::read(env!("CARGO_MANIFEST_DIR").to_owned() + "/certs/ca-cert.pem").unwrap();
+    config.set_security_policy(&DEFAULT_TLS13).unwrap();
+    config.trust_pem(&ca).unwrap();
+    config.build().unwrap()
+}
+
+#[test]
+fn scenario() -> turmoil::Result {
+    // turmoil is a network simulator, so we can simulate running a single and
+    // two servers without having to spin up multiple processes or wait for
+    // real time to elapse
+    let mut sim = turmoil::Builder::new().build();
```

Comment:
Do we really need turmoil here? I'd prefer to have examples a bit more focused and only related to what we're trying to show. Otherwise, the reader needs to do more work to split apart the setup here.

That being said, I think it'd be great to have a separate turmoil example.

Reply:
I don't think we strictly need turmoil here, but it was a complex enough example that I wanted some form of tests, which turmoil made extremely convenient, and it also kept everything in the same file which I hoped would make up for the hit in readability. 

If we're okay with just manual test for the examples then I'll go ahead and move it to a process oriented model.